File: Graphs/0417. Pacific Atlantic Water Flow.cs
public class Solution {
    private int rows = 0, cols = 0;
    private int[][] heights;

    public IList<IList<int>> PacificAtlantic(int[][] heights) {
        rows = heights.Length;
        cols = heights[0].Length;
        this.heights = heights;
        var pacific = new HashSet<(int X, int Y)>();
        var atlantic = new HashSet<(int X, int Y)>();

        for(int i = 0; i < cols; i++){ // Iterate through each column
            DFS(0, i, heights[0][i], pacific);
            DFS(rows - 1, i, heights[rows - 1][i], atlantic);
        }

        for(int i = 0; i < rows; i++){ // Iterate through each row
            DFS(i, 0, heights[i][0], pacific);
            DFS(i, cols - 1, heights[i][cols - 1], atlantic);
        }

        var res = new List<IList<int>>();
        for(int x = 0; x < rows; x++){ // Brute Force check if given cell can reach both oceans
            for(int y = 0; y < cols; y++){
                (int X, int Y) cur = (x, y);
                if(pacific.Contains(cur) && atlantic.Contains(cur)){
                    res.Add(new List<int>() {x, y});
                }
            }
        }

        return res;
    }

    private void DFS(int x, int y, int prevHeight, HashSet<(int X, int Y)> ocean){
        (int X, int Y) cur = (x, y);
        if(ocean.Contains(cur) || x < 0 || x >= rows || y < 0 || y >= cols ||
         prevHeight > heights[x][y]){ // current height is lower
            return; // Out of bands or already visited
        }

        ocean.Add(cur);

        DFS(x + 1, y, heights[x][y], ocean);
        DFS(x - 1, y, heights[x][y], ocean);
        DFS(x, y + 1, heights[x][y], ocean);
        DFS(x, y - 1, heights[x][y], ocean);
    }
}

File: Graphs/0417. Pacific Atlantic Water Flow.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        ROWS, COLS = len(heights), len(heights[0])
        pac, atl = set(), set()
        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]

        def dfs(r, c, visit, prevHeight):
            if ((r, c) in visit or 
                r < 0 or c < 0 or r == ROWS or c == COLS or
                heights[r][c] < prevHeight):
                return
            
            visit.add((r, c))
            for dr, dc in directions:
                ro, co = r + dr, c + dc
                dfs(ro, co, visit, heights[r][c])


        for c in range(COLS):
            dfs(0, c, pac, heights[0][c])
            dfs(ROWS - 1, c, atl, heights[ROWS - 1][c])

        for r in range(ROWS):
            dfs(r, 0, pac, heights[r][0])
            dfs(r, COLS - 1, atl, heights[r][COLS - 1])

        res = []
        for r in range(ROWS):
            for c in range(COLS):
                if (r, c) in pac and (r, c) in atl:
                    res.append([r, c])
        return res
